// ionic4-date-picker/__ivy_ngcc__/dist/date-picker.module.js
import * as ɵngcc0 from '@angular/core';
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { NgModule } from '@angular/core';
import { IonicModule } from '@ionic/angular';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { DatePickerComponent } from './components/date-picker-component/date-picker-component';
var DatePickerModule = /** @class */ (function () {
    function DatePickerModule() {
    }
    DatePickerModule_1 = DatePickerModule;
    DatePickerModule.forRoot = function () {
        return {
            ngModule: DatePickerModule_1,
            providers: []
        };
    };
    var DatePickerModule_1;
DatePickerModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: DatePickerModule });
DatePickerModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function DatePickerModule_Factory(t) { return new (t || DatePickerModule)(); }, imports: [[
            IonicModule,
            CommonModule,
            FormsModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(DatePickerModule, { declarations: function () { return [DatePickerComponent]; }, imports: function () { return [IonicModule,
        CommonModule,
        FormsModule]; }, exports: function () { return [DatePickerComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(DatePickerModule, [{
        type: NgModule,
        args: [{
                imports: [
                    IonicModule,
                    CommonModule,
                    FormsModule
                ],
                declarations: [
                    DatePickerComponent,
                ],
                exports: [
                    DatePickerComponent
                ]
            }]
    }], function () { return []; }, null); })();
    return DatePickerModule;
}());
export { DatePickerModule };
