// ng2-canvas-whiteboard/esm2015/lib/shapes/canvas-whiteboard-shape-selector.component.js
import { Component, ElementRef, EventEmitter, Input, Output } from '@angular/core';
import { CanvasWhiteboardShapeService } from './canvas-whiteboard-shape.service';
export class CanvasWhiteboardShapeSelectorComponent {
    constructor(elementRef, canvasWhiteboardShapeService) {
        this.elementRef = elementRef;
        this.canvasWhiteboardShapeService = canvasWhiteboardShapeService;
        this.showShapeSelector = false;
        this.onToggleShapeSelector = new EventEmitter();
        this.onShapeSelected = new EventEmitter();
        this.registeredShapes$ = this.canvasWhiteboardShapeService.registeredShapes$;
    }
    selectShape(shape) {
        this.onShapeSelected.emit(shape);
        this.toggleShapeSelector(null);
    }
    closeOnExternalClick(event) {
        if (!this.elementRef.nativeElement.contains(event.target) && this.showShapeSelector) {
            this.onToggleShapeSelector.emit(false);
        }
    }
    toggleShapeSelector(event) {
        if (event) {
            event.preventDefault();
        }
        this.onToggleShapeSelector.emit(!this.showShapeSelector);
    }
}
CanvasWhiteboardShapeSelectorComponent.decorators = [
    { type: Component, args: [{
                selector: 'canvas-whiteboard-shape-selector',
                host: {
                    '(document:mousedown)': 'closeOnExternalClick($event)',
                    '(document:touchstart)': 'closeOnExternalClick($event)'
                },
                template: `
    <div *ngIf="!showShapeSelector" (click)="toggleShapeSelector($event)"
         class="canvas-whiteboard-shape-selector-selected-preview">
      <canvas-whiteboard-shape-preview [shapeConstructor]="selectedShapeConstructor"
                                       [shapeOptions]="shapeOptions"></canvas-whiteboard-shape-preview>
    </div>
    <div class="canvas-whiteboard-shape-selector-wrapper" *ngIf="showShapeSelector">
      <canvas-whiteboard-shape-preview *ngFor="let shapeConstructor of registeredShapes$ | async"
                                       [shapeConstructor]="shapeConstructor"
                                       [shapeOptions]="shapeOptions"
                                       (click)="selectShape(shapeConstructor)"></canvas-whiteboard-shape-preview>
    </div>
  `,
                styles: [`
    .canvas-whiteboard-shape-selector-selected-preview {
      vertical-align: bottom;
      display: inline-block;
    }

    .canvas-whiteboard-shape-selector-wrapper {
      display: block;
      padding: 4px;
      border: 1px solid #afafaf;
    }

    @media (min-width: 401px) {
      .canvas-whiteboard-shape-selector-wrapper {
      }
    }
  `]
            },] }
];
CanvasWhiteboardShapeSelectorComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: CanvasWhiteboardShapeService }
];
CanvasWhiteboardShapeSelectorComponent.propDecorators = {
    showShapeSelector: [{ type: Input }],
    selectedShapeConstructor: [{ type: Input }],
    shapeOptions: [{ type: Input }],
    onToggleShapeSelector: [{ type: Output }],
    onShapeSelected: [{ type: Output }]
};
