// ng2-canvas-whiteboard/esm2015/lib/canvas-whiteboard.service.js
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
export class CanvasWhiteboardService {
    constructor() {
        this._canvasDrawSubject = new Subject();
        this.canvasDrawSubject$ = this._canvasDrawSubject.asObservable();
        this._canvasClearSubject = new Subject();
        this.canvasClearSubject$ = this._canvasClearSubject.asObservable();
        this._canvasUndoSubject = new Subject();
        this.canvasUndoSubject$ = this._canvasUndoSubject.asObservable();
        this._canvasRedoSubject = new Subject();
        this.canvasRedoSubject$ = this._canvasRedoSubject.asObservable();
    }
    drawCanvas(updates) {
        this._canvasDrawSubject.next(updates);
    }
    clearCanvas() {
        this._canvasClearSubject.next();
    }
    undoCanvas(updateUUD) {
        this._canvasUndoSubject.next(updateUUD);
    }
    redoCanvas(updateUUD) {
        this._canvasRedoSubject.next(updateUUD);
    }
}
CanvasWhiteboardService.decorators = [
    { type: Injectable }
];
